### ISSUE TITLE
HTTP API 'show routes' capture

### ISSUE THREAD

**Issue by switchtower** [Exa-Networks/exabgp#326]
First of all, I've been following ExaBGP for a while now and recently started replacing our 'trigger' routers with ExaBGP for RTBH and internal null routes.  It's worked well so far, but I'm having problems getting the routes from Exa to compare with what is currently in the master DB: there are multiple servers running in each ASN that are kept up-to-date via message queues when people add/remove/modify the routes via an HTTP form.  I periodically want to compare the master DB with what Exa has and make sure the correct IP addresses are being being blocked.

The way the routes are initially being added to the config is I start the daemon up and 'bulk' add via an 'announce attribute' and have all the nulls in the NLRI.  At that point, when something is received, via the message queue, I post to the HTTP API I have running, which also works great.

The problem comes when I want to grab the routes and compare them to the DB.  I have a 'dirty' script to grab the routes:

```
def get_routes(peer):
    post('http://127.0.0.1:{0}'.format(peer.port), {'command': 'show routes'})
    time.sleep(10)
    lines = []
    routes = []
    for line in reversed(open(os.path.join(settings.BASE_DIR, "exabgp_{0}.log".format(peer.port))).readlines()):
        if 'show routes' in line:
            break
        else: lines.append(line)
    for line in lines:
        if 'Responding to http_api' in line:
            routes.append(line.split()[9])
    return routes
```
I couldn't figure out a more elegant way of grabbing the returned routes from the sys.stdout in the post, so I'm grabbing them from the logs the best I can. (any suggestions here would be great!)

When I do this, I only get about 1110 returned routes, and the BGP session I have with my Cisco router dies and I'm then unable to gracefully kill the exabgp process:
```
Aug 27 15:26:56: %BGP-5-ADJCHANGE: neighbor 172.16.130.57 Down BGP Notification sent
Aug 27 15:26:56: %BGP-3-NOTIFICATION: sent to neighbor 172.16.130.57 4/0 (hold time expired) 0 bytes
Aug 27 15:26:56: %BGP_SESSION-5-ADJCHANGE: neighbor 172.16.130.57 IPv4 Unicast topology base removed from session  BGP Notification sent
```
```
exabgp 9477   processes     Error received while SENDING data to helper program, retrying ([Errno EINTR] [Errno 4] Interrupted system call)
exabgp 9477   reactor       SIG TERM received - shutdown
```
My config:
```
group noc_trigger {
        router-id 172.16.130.57;
        graceful-restart 1200;
        hold-time 90;
        local-as 65535;

        capability {
            route-refresh;
        }

        process logtodb {
                run /home/noc/noc_trigger/bin/python /home/noc/noc_trigger/noc_trigger/manage.py logtodb;
                encoder json;
                neighbor-changes;

        }

        process http_api {
                run /home/noc/noc_trigger/bin/python /home/noc/noc_trigger/noc_trigger/manage.py http_api 5001;
        }
        neighbor 172.16.128.16 {
                local-address 172.16.130.57;
                peer-as 65535;
                local-as 65535;
                md5 57y3l1bCLTNNLNypn6;
                group-updates;
        }

}
```

Any help anyone can provide would be wonderful. :)

**Comment by thomas-mangin** [comment]
Hello @switchtower,

which version or ExaBGP are you using, older version of ExaBGP would not handle neighbor messages while performing the dump while newer should.

Thomas

**Comment by switchtower** [comment]
```
ExaBGP : 3.4.12
Python : 2.7.5 (default, Jun 24 2015, 00:41:19)  [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]
Uname  : #1 SMP Fri Mar 6 11:36:42 UTC 2015
```

**Comment by thomas-mangin** [comment]
:sob: 
Fixing this on 3.4 would most likely be a major work I would rather not start if possible.
Would you mind using the latest master ? AFAICR it is fixed already there.

**Comment by switchtower** [comment]
I'll update to master and let you know what I find.  Thanks for the quick responses! :)

**Comment by switchtower** [comment]
updated to 3.9, but it doesn't appear to like the 3.4 configuration file:
```
Fri, 28 Aug 2015 11:28:13 | INFO     | 22880  | reactor       | performing reload of exabgp 3.9.0
Fri, 28 Aug 2015 11:28:13 | INFO     | 22880  | configuration | > group            | 'noc_trigger'
Fri, 28 Aug 2015 11:28:13 | ERR      | 22880  | configuration | problem with the configuration file, no change done
Fri, 28 Aug 2015 11:28:13 | ERR      | 22880  | configuration |
```
I'll keep digging, again, thanks for your help. :)

**Comment by thomas-mangin** [comment]
Please look at the ./qa folder for example of the new format ( the JSON format was also slightly modified, changes are listed in the CHANGELOG file )

**Comment by switchtower** [comment]
Getting around to testing this, I'm still seeing similar results running 3.9:
```
ExaBGP : 3.9.0
Python : 2.7.5 (default, Jun 24 2015, 00:41:19)  [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]
Uname  : Linux lw-dc5-trigger1.noc.switchtower.com 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015 x86_64
```
```
In [9]: r = get_routes(p)

In [10]: len(r)
Out[10]: 1126
```
Total number of routes should be around 4126.

The logs stop and the peer dies:
```
lw-dc5-core1.rtr.switchtower.com :: sent to neighbor 185.67.44.200 4/0 hold time expired 0 bytes
lw-dc5-core1.rtr.switchtower.com :: neighbor 185.67.44.200 Down BGP Notification sent
lw-dc5-core1.rtr.switchtower.com :: neighbor 185.67.44.200 IPv4 Unicast topology base removed from session  BGP Notification sent
lw-dc5-core1.rtr.switchtower.com :: Neighbor 185.67.44.200 reset BGP Notification sent
```
Configuration:
```
process logtodb {
        run /home/noc/noc_trigger/bin/python /home/noc/noc_trigger/noc_trigger/manage.py logtodb;
        encoder json;
}

process http_api {
        run /home/noc/noc_trigger/bin/python /home/noc/noc_trigger/noc_trigger/manage.py http_api 5002;
        encoder json;
}

neighbor 172.16.128.16 {
    local-address 172.16.130.57;
    router-id 172.16.130.57;
    hold-time 90;
    peer-as 65535;
    local-as 65535;

    capability {
        graceful-restart;
    }

    static {
        route 2.56.0.0/14 next-hop 192.0.2.1 community [ 65535:65281 ] local-preference 200;
        route 5.8.248.0/21 next-hop 192.0.2.1 community [ 65535:65281 ] local-preference 200;
        route 5.45.32.0/20 next-hop 192.0.2.1 community [ 65535:65281 ] local-preference 200;
    }

    api {
        processes [ logtodb, http_api ];
    }

}
```

**Comment by thomas-mangin** [comment]
Hi @switchtower - thank you for the extra information.

It is still hard to see what could cause this.

I am surprised by the syntax you are providing ( using a comma between each process name ) I do not have a test suite for this case yet. It may be valid syntax but I would not recommend using any comma in the list.

"announce attributes" is useful to reduce parsing on ExaBGP for announcing large number of routes, it is great to see that you are already using it. I intend to provide named attribute at some point (do not hold your breath tho).

I have not hacked on ExaBGP for a little but AFAICR the show routes command should not be blocking in the master branch (3.9). Are your telling me that the show routes command is causing the BGP session to drop ?

Are you sure you are parsing the UPDATE message correctly as it may be useful to check that you do not see 1126 UPDATEs vs NLRIs ?

The log line I assume you are looking for is the one generated by 
```
self.logger.routes(LazyFormat(self.me('<< UPDATE (%d) nlri ' % number),nlri,str))
```

Without looking at the code of the HTTP server, it is hard to guess what is happening under the hood and could cause this issue for you as it is hard to understand your code snippet out of context ... I do not know what debug option you have enabled on ExaBGP which would also affect the log output.

**Comment by thomas-mangin** [comment]
I have just pushed a fix for #337. It could cause some data to be sent to the wrong process (and therefore cause some blocking I/O as it would not be consumed). Could you please update to the latest 3.4 branch and please tell me if the issue is still there.

**Comment by thomas-mangin** [comment]
I will assume that this issue was fixed by the patch provided - if not, feel free to re-open.
